[cssom-view] Implement Element.prototype.checkVisibility API

nt1m opened the pull request:
#### 7d04a9c1b62bc8a8101c390f6a164cf4b3da9cfa
<pre>
[cssom-view] Implement Element.prototype.checkVisibility API
<a href="https://bugs.webkit.org/show_bug.cgi?id=263223">https://bugs.webkit.org/show_bug.cgi?id=263223</a>
rdar://117113659

Reviewed by Simon Fraser.

Implement the API defined at: <a href="https://drafts.csswg.org/cssom-view/#dom-element-checkvisibility">https://drafts.csswg.org/cssom-view/#dom-element-checkvisibility</a>

This is implemented in a way that exclusively queries DOM + style information, to report correct information on skipped subtrees
without performing a forced layout on those subtrees.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/checkVisibility-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/checkVisibility.html: Resync from upstream
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/CheckVisibilityOptions.h: Added.
* Source/WebCore/dom/CheckVisibilityOptions.idl: Copied from Source/WebCore/dom/Element+CSSOMView.idl.
* Source/WebCore/dom/Element+CSSOMView.idl:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::checkVisibility):
* Source/WebCore/dom/Element.h:

Canonical link: <a href="https://commits.webkit.org/269506@main">https://commits.webkit.org/269506@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3098255ae1bc80942397412e2ddef86caf7c9555

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22784 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/600 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23871 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24692 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21089 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23041 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/1521 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23310 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23024 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19760 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25545 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20640 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/26850 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/19836 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/20644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20884 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24706 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/22146 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/328 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/18154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/26203 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20443 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6138 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5424 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/27485 "Built successfully") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5967 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5421,6 +5422,41 @@ void Element::setContentRelevancy(OptionSet<ContentRelevancy> contentRelevancy)
+    if (options.checkVisibilityCSS && style->visibility() != Visibility::Visible)
+        return false;
+
+    for (RefPtr ancestor = this; ancestor; ancestor = ancestor->parentElementInComposedTree()) {
```
The spec says to check the shadow-including ancestors of "this", not "in flat tree". They're different. In particular, a node assigned to a slot will have its parent node as the parent in shadow-including ancestors but will have its assigned slot as the part in flat tree.

nt1m replied:
I think the flat tree is correct here, since we want to match CSS rendering. This is in fact what Chrome implements. I'll file a CSSWG issue to get this corrected.

rniwa replied:
We should definitely add a test for this.

nt1m replied:
I filed https://github.com/w3c/csswg-drafts/issues/9486 , I'll add a test tomorrow

rniwa commented on Source/WebCore/dom/CheckVisibilityOptions.idl:
```diff
@@ -0,0 +1,30 @@
+// https://drafts.csswg.org/cssom-view/#extension-to-the-element-interface
+dictionary CheckVisibilityOptions {
+    boolean checkOpacity = false;
+    boolean checkVisibilityCSS = false;
```
It looks inconsistent that the one of the options is called checkOpacity and the other one is called checkVisibilityCSS. Why don't both of them have CSS suffix?

nt1m replied:
Yeah this isn't particularly great. The CSS suffix was added to disambiguate from the method name, but it is inconsistent. We should probably get this spec changed in some way.

nt1m replied:
https://github.com/w3c/csswg-drafts/issues/9487

nt1m replied:
We should rename in a follow up patch when the CSSWG resolves

rniwa commented on Source/WebCore/dom/Element+CSSOMView.idl:
```diff
@@ -27,6 +27,9 @@
     DOMRectList getClientRects();
     [NewObject] DOMRect getBoundingClientRect();
+
+    boolean checkVisibility(optional CheckVisibilityOptions options = {});
```
We should put this under a runtime flag at least.

nt1m replied:
I can add a flag to disable, but we'll want it on by default so reader can make use of this API.

rniwa replied:
We have the ability to expose API to a specific DOM wrapper world. See `Element::shadowRootForBindings` and `InjectedBundleScriptWorld::makeAllShadowRootsOpen()` and `collectMatchingElementsInFlatTree` and `matchingElementInFlatTree` in `LocalDOMWindow.idl`.

smfr commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5421,6 +5422,41 @@ void Element::setContentRelevancy(OptionSet<ContentRelevancy> contentRelevancy)
+    }
+
+    // skippedContentReason() includes the skipped content root, so we query the parent to make sure we exclude roots that are visible.
+    auto* parent = parentElementInComposedTree();
```
This part isn't in the spec yet, right? I think you should land initially without the content-visibility part.

nt1m replied:
It is part of the spec. This is excluding c-v: hidden. The bit not in the spec yet is c-v: auto

nt1m committed:
[cssom-view] Implement Element.prototype.checkVisibility API
https://bugs.webkit.org/show_bug.cgi?id=263223
rdar://117113659

Reviewed by Simon Fraser.

Implement the API defined at: https://drafts.csswg.org/cssom-view/#dom-element-checkvisibility

This is implemented in a way that exclusively queries DOM + style information, to report correct information on skipped subtrees
without performing a forced layout on those subtrees.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/checkVisibility-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/checkVisibility.html: Resync from upstream
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/CheckVisibilityOptions.h: Added.
* Source/WebCore/dom/CheckVisibilityOptions.idl: Copied from Source/WebCore/dom/Element+CSSOMView.idl.
* Source/WebCore/dom/Element+CSSOMView.idl:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::checkVisibility):
* Source/WebCore/dom/Element.h:

Canonical link: https://commits.webkit.org/269506@main
